### ISSUE TITLE
Running test suite from Test Explorer calls code in other test suites

### ISSUE THREAD

**Issue by apkerr** [hbenl/vscode-mocha-test-adapter#28]
I'm not sure if this is a test explorer thing, or a mocha-test-adapter thing, so let me know if it should go in another place. I've simplified the issue down to the following:

setup.js
```javascript
module.exports = {
  init: function(msg) {
    console.log(msg);
  }
};
```
test1_ts.js
```javascript
var setup = require("./setup");

describe("Test Suite 1", function() {
    setup.init("Initializing Test Suite 1");
    it("Test 1", function() {
      console.log("Running Test Suite 1, test 1");
    });
});
```
test2_ts.js
```javascript
var setup = require("./setup");

describe("Test Suite 2", function() {
  setup.init("Initializing Test Suite 2");
  it("Test 1", function() {
    console.log("Running Test Suite 2, test 1");
  });
});
```
The test explorer discovers and shows the tests properly:
![test_explorer](https://user-images.githubusercontent.com/23730598/50244287-93724880-039d-11e9-9231-a81ced53856f.png)

When I run mocha manually on test1_ts, I get the expected output:
```
$ mocha test1_ts
Initializing Test Suite 1

  Test Suite 1
Running Test Suite 1, test 1
    ✓ Test 1

  1 passing (6ms)
```
When I run through the test explorer though, I can see that test suite 2 was also initialized:
Output->Mocha Tests
```
Initializing Test Suite 2
Initializing Test Suite 1
Running Test Suite 1, test 1
```
The Mocha Explorer Log has the following:
```
[2018-12-19 19:52:02.375] [INFO] Running test(s) ["/home/andrew/analytics/analytics-app/test/test1_ts.js: Test Suite 1 Test 1"] of /home/andrew/analytics/analytics-app
[2018-12-19 19:52:02.376] [DEBUG] Looking for test files test/**/*.js in /home/andrew/analytics/analytics-app
[2018-12-19 19:52:02.444] [DEBUG] Found test files ["/home/andrew/analytics/analytics-app/test/test2_ts.js","[workspace path]/test/test1_ts.js","/home/andrew/analytics/analytics-app/test/setup.js"]
[2018-12-19 19:52:02.456] [DEBUG] Using nodePath: /usr/bin/node
[2018-12-19 19:52:02.457] [DEBUG] Looking for mocha options in /home/andrew/analytics/analytics-app/test/mocha.opts
[2018-12-19 19:52:02.457] [DEBUG] Options from mocha.opts file: {"requires":[]}
[2018-12-19 19:52:02.458] [DEBUG] Using Mocha options: {"ui":"bdd","timeout":20000,"retries":0,"requires":[],"exit":true}
[2018-12-19 19:52:02.458] [DEBUG] Using working directory: /home/andrew/analytics/analytics-app
[2018-12-19 19:52:02.458] [DEBUG] Using environment variable config: {}
[2018-12-19 19:52:02.607] [INFO] Worker: Running tests
[2018-12-19 19:52:02.616] [INFO] Received {"type":"suite","suite":"undefined: ","state":"running"}
[2018-12-19 19:52:02.619] [INFO] Received {"type":"suite","suite":"/home/andrew/analytics/analytics-app/test/test1_ts.js: Test Suite 1","state":"running"}
[2018-12-19 19:52:02.619] [INFO] Received {"type":"test","test":"/home/andrew/analytics/analytics-app/test/test1_ts.js: Test Suite 1 Test 1","state":"running"}
[2018-12-19 19:52:02.620] [INFO] Received {"type":"test","test":"/home/andrew/analytics/analytics-app/test/test1_ts.js: Test Suite 1 Test 1","state":"passed"}
[2018-12-19 19:52:02.621] [INFO] Received {"type":"suite","suite":"/home/andrew/analytics/analytics-app/test/test1_ts.js: Test Suite 1","state":"completed"}
[2018-12-19 19:52:02.621] [INFO] Received {"type":"suite","suite":"undefined: ","state":"completed"}
[2018-12-19 19:52:02.623] [INFO] Worker finished
```

Throwing a breakpoint in setup.js and running debug on Test 1, I can see the following process running:
```
$ ps -ef | grep mocha
andrew   22158 16905  9 14:58 ?        00:00:00 /usr/bin/node --inspect-brk=9229 /home/andrew/.vscode/extensions/hbenl.vscode-mocha-test-adapter-1.5.1/out/worker/runTests.js ["/home/andrew/analytics/analytics-app/test/test2_ts.js","/home/andrew/analytics/analytics-app/test/test1_ts.js","/home/andrew/analytics/analytics-app/test/setup.js"] ["Test Suite 1 Test 1"] "/home/andrew/.vscode/extensions/hbenl.vscode-mocha-test-adapter-1.5.1/node_modules/mocha/index.js" {"ui":"bdd","timeout":20000,"retries":0,"requires":[],"exit":true} true
```
In the debugger I can see that the call to init is coming from test2_ts.js, even though I only want to run test1_ts.js:
```
init (/home/andrew/analytics/analytics-app/test/setup.js:3)
(anonymous function) (/home/andrew/analytics/analytics-app/test/test2_ts.js:4)
create (/home/andrew/.vscode/extensions/hbenl.vscode-mocha-test-adapter-1.5.1/node_modules/mocha/lib/interfaces/common.js:112)
context.describe.context.context (/home/andrew/.vscode/extensions/hbenl.vscode-mocha-test-adapter-1.5.1/node_modules/mocha/lib/interfaces/bdd.js:40)
(anonymous function) (/home/andrew/analytics/analytics-app/test/test2_ts.js:3)
Module._compile (module.js:652)
Module._extensions..js (module.js:663)
Module.load (module.js:565)
tryModuleLoad (module.js:505)
Module._load (module.js:497)
Module.require (module.js:596)
require (module.js:11)
(anonymous function) (/home/andrew/.vscode/extensions/hbenl.vscode-mocha-test-adapter-1.5.1/node_modules/mocha/lib/mocha.js:250)
Mocha.loadFiles (/home/andrew/.vscode/extensions/hbenl.vscode-mocha-test-adapter-1.5.1/node_modules/mocha/lib/mocha.js:247)
Mocha.run (/home/andrew/.vscode/extensions/hbenl.vscode-mocha-test-adapter-1.5.1/node_modules/mocha/lib/mocha.js:576)
(anonymous function) (/home/andrew/.vscode/extensions/hbenl.vscode-mocha-test-adapter-1.5.1/out/worker/runTests.js:39)
Module._compile (module.js:649)
Module._extensions..js (module.js:663)
Module.load (module.js:565)
tryModuleLoad (module.js:505)
```

**Comment by apkerr** [comment]
I can see that the test-explorer passes the list of all the files found in the mochaExplorer.files glob to the mocha-test-adapter.

The test adapter then passes all those files to mocha. Mocha then loads all the files, which executes code that is in the ``describe`` block before the first ``it``. It then runs the selected test.

It looks like the issue is that the file list sent to mocha contains all the files in the glob, regardless of whether or not those files are needed in order to run the selected test. If I trim down my glob to only show test 1 in the test-explorer, only Test Suite 1 is initialized. That is the same behaviour as executing the test file from the command line.

I am relatively inexperienced with js/mocha testing, but have recently volunteered to take over our web automation and have a pretty large pile of tests to sort through and optimize. So it's quite possible that I am doing something wrong or that our tests are not arranged in the 'usual' way, whatever that is. Please let me know if I'm misunderstanding something.

**Comment by uniquejava** [comment]
I met similar issue. I am test a simple `hello.spec.js` in a big node.js project.

This hello.spec.js just tests something like 1+1 equals 2 or not .

However, running this single file, seems bring up my whole node.js app. 
Looks like`npm start` in my package.json is called before each test. 

I am not sure a single hello world test will start my whole node.js app.

Any clue?

The following are logs (from Mocha Test console).
```
Error: listen EADDRINUSE :::10010
    at Server.setupListenHandle [as _listen2] (net.js:1360:14)
    at listenInCluster (net.js:1401:12)
    at Server.listen (net.js:1485:7)
    at Function.listen (BookService/app/node_modules/express/lib/application.js:618:24)
    at /Users/xxxx/Documents/Training/vbook/BookService/app/app.js:35:9
    at /Users/xxxx/Documents/Training/vbook/BookService/app/node_modules/swagger-express-mw/lib/index.js:32:5
    at /Users/xxxx/Documents/Training/vbook/BookService/app/node_modules/swagger-node-runner/index.js:234:7
    at <anonymous>
```

**Comment by uniquejava** [comment]
Thank you. @apkerr 
>It looks like the issue is that the file list sent to mocha contains all the files in the glob, regardless of whether or not those files are needed in order to run the selected test. If I trim down my glob to only show test 1 in the test-explorer, only Test Suite 1 is initialized. That is the same behaviour as executing the test file from the command line.

Based on your findings, I changed my settings 
from`"mochaExplorer.files": "BookService/app/test/**/*.js"` 
to `"mochaExplorer.files": "BookService/app/test/unit/**/*.js"`

which excludes a `controller` directory, in that directory, a spec which has this line `const app = require('../app')` which this plugin runs and which starts my node.js app (even it's not selected for testing)

Now my tests run super fast. 

But yes, still anyone knows how to prevent this vscode-mocha-test-adapter from calling code in other test suites?

**Comment by hbenl** [comment]
I have added the `mochaExplorer.pruneFiles` config option to the latest release (1.6.0) of this adapter. If you set this option to `true`, only the files containing the tests selected in the current test run are loaded. The default behavior is still to load all files found in the glob: some people put setup code that is needed by all tests into separate files and load them through the glob - in this case, `mochaExplorer.pruneFiles` would break the tests.

**Comment by apkerr** [comment]
Thanks. A quick test shows that it's behaving the way I expected with the flag set to true.

**Comment by humbienri** [comment]
Please include this in the primary documentation for this plugin. Please! This has been quite a source of time spent.

So it seems that it is not possible to Run/Debug a single test case/spec or a single suite?

**Comment by hbenl** [comment]
@humbienri Of course you should be able to Run/Debug a single test or suite. However, without the `pruneFiles` option, all test files will be executed first to register all tests before the selected test(s) will be run (executing a test file does not execute the tests contained in it, it only registers the tests with Mocha) - this is standard Mocha behavior.
This only becomes a problem if a test file is written so that executing it has side-effects. Test files should not have side-effects, if you need to run some setup routine before the tests in a file are run, then you should use Mocha's `before()` [hook](https://mochajs.org/#hooks). This will register your setup routine with Mocha and it will only be executed by Mocha if it is needed by one of the selected tests.

**Comment by humbienri** [comment]
Again, I apologize @hbenl. I barely now am seeing your comment from 14 days ago.  I did manage to figure out how to Debug a single test. I think that at the time I wrote my comment, I had some .only sprinkled somewhere that was interfering with the Test Explorer.  I think that is what was happening but I am not certain.  I inherited these suites of tests and javascript is really not my forte so I'll have to dig into what is happening as it pertains to "side effects".  Thank you!
